Feature/8s - Improve Babel config

bebraw opened the pull request:
---
Closes #8
---

**How to test (feature)**

Good question. Should I set up a small webpack build to this project to test?

**How to test (potential regressions)**

bebraw committed:
chore: Set up Babel lodash plugin

bebraw committed:
chore: Add suggested Babel configuration

bebraw committed:
chore: Add a changeset

TkDodo commented on package.json:
```diff
@@ -54,6 +54,7 @@
     "@types/brotli-webpack-plugin": "^1.1.0",
     "app-root-path": "3.0.0",
     "babel-loader": "^8.1.0",
+    "babel-plugin-lodash": "3.3.4",
```
```suggestion
    "app-root-path": "^3.0.0",
    "babel-loader": "^8.1.0",
    "babel-plugin-lodash": "^3.3.4",
```
I know I've said the opposite quite some time, but for `libraries`, we actually want "compatible with" versions so that we avoid duplication of dependencies if we require exact versions (e.g. 3.3.4 from here and maybe 3.3.2 from somewhere else, we'll wind up with 2 versions). If we use `compatible with`, it will just use 3.3.4 for both of them.

for our products, we want exact versions though because we have package-lock anyways and it makes it easier to see what is really installed.

TkDodo commented on .babelrc:
```diff
@@ -16,6 +20,9 @@
+      "plugins": [["lodash", { "id": ["lodash"] }]]
+    },
     "test": {
       "presets": ["@babel/preset-env", "@babel/preset-react"],
```
do we need this? we have both presets in the global config as well, this looks like it just resets preset-env to the default for `test`. do you know why we have it?

bebraw replied:
Likely it came to the project through the initial work and we forgot to remove `@babel/preset-env` from here. Let's do that now. 👍

bebraw replied:
It's the same with React preset as well.

bebraw replied:
Ok, changed.

TkDodo committed:
Update package.json

use compatible-with versions for libraries

bebraw committed:
chore: Drop presets from the test target at Babel

Those bits seem redundant now.

TkDodo commented on .changeset/8s-improve-babel-config.md:
```diff
@@ -0,0 +1,20 @@
+---
+issue: https://github.com/pmedianetwork/webpack-config/issues/8
+label: Enhancements
+---
```
> How to mark this as a breaking release?

@reakaleek I think we just need todo `type: major` instead of the `label`, right?

reakaleek replied:
yes `type: major` will do


You only need a `label`  If you can't provide a type (e.g. documentation changes, or build process changes)
and for "normal" situations you need to choose a `type`  whereas `type` is one of `major`, `minor` or `patch`

reakaleek replied:
btw... from my understanding 0.x.x releases are always considered as unstable and breaking. I would only bump to 1.0.0.. if you think this is 1.0.0 worthy...

bebraw replied:
Ok, thanks for the info. Since we're still in `0.x` series, I'll change to `type: minor` to follow SemVer. Eventually we should do 1.0 and do breaking changes with `type: major`.

reakaleek replied:
I've just reread semver 2.0.. and it feels like this should be 1.0.0 already :D


>How should I deal with revisions in the 0.y.z initial development phase?
The simplest thing to do is start your initial development release at 0.1.0 and then increment the minor version for each subsequent release.
>
>How do I know when to release 1.0.0?
If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

bebraw replied:
Technically we could call this one as 1.0.0. 👍

bebraw replied:
Ok, I marked this one with `type: major`. It's going to be 1.0.

Before merging, it might be a good idea to look into testing the change against a small local setup so we have means to compare output without having to test against the full projects.

reakaleek replied:
can we set this up as some kind of integration test? in a subfolder?

reakaleek replied:
probably needs some symlinking to make it work hmmmmfff

bebraw replied:
Yeah, that's what I was thinking. I would do a directory with a small webpack configuration that's running using the code. Technically we could treat the output as snapshots and make sure the build files get updates so they can be reviewed on each PR or so.

reakaleek replied:
Do want to do this in this PR or a follow up?

bebraw replied:
I would do a quick setup here so we have a good baseline to compare against. It's not much work to add.

reakaleek replied:
Okay nice.. let me know if you need help with the pipeline

bebraw replied:
Sure. 👍

bebraw committed:
chore: Mark as a minor release

bebraw committed:
chore: Mark as a major release

This will be 1.0.

bebraw committed:
chore: Set up a basic webpack configuration for testing

The next step is to wire this with the project and add a simple snapshot
to compare against.

bebraw committed:
chore: Sketch out snapshotting

Only mem fs needs to be resolved. The question is how to get the data
from there.

bebraw committed:
chore: Write a snapshot

Now it captures file output from the build and tests against it.

bebraw committed:
fix: Make sure test dependencies are installed before testing

bebraw commented:
I have a basic snapshot test in place. It's writing a webpack build against a memory FS and then it's asserting that so we can see if something has changed in the build because of a configuration change.

The webpack configuration that's in place is somewhat simple and so is the code. Ideally we would add especially code that matches real world usage to see how it transpiles.

@reakaleek Can you paste the error from the CI so I can debug?

TkDodo commented:
seems like we're missing `npm install`:

```
> @pmedianetwork/webpack-config@0.9.4 test /home/ubuntu/jenkins-slave/workspace/webpack-config/pull-request-test
> jest test.js

sh: 1: jest: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! @pmedianetwork/webpack-config@0.9.4 test: `jest test.js`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the @pmedianetwork/webpack-config@0.9.4 test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?
```

bebraw committed:
fix: Make sure Jest is installed

bebraw committed:
chore: Push the npm install bit to jenkins

That's a better place.

bebraw commented:
@TkDodo Ok, thanks!

It's good for a look. Most likely the test can be improved. The question is more about what can we break there with the Babel change. One option is to improve the test later now that the hard part is done.

reakaleek commented on .jenkins/test.Jenkinsfile:
```diff
@@ -8,7 +8,7 @@ pipeline {
             steps {
                 nvm(env.NODE_VERSION) {
-                    sh 'npm run test'
+                    sh 'npm install && npm run test'
```
Please use `npm ci`

bebraw replied:
Sure, good catch.

bebraw committed:
fix: Use npm ci

bebraw committed:
fix: Use npm ci

TkDodo commented:
> One option is to improve the test later now that the hard part is done.

yes, we can have improve the test later, also maybe add some es6 code to see if it's preserved etc ...

bebraw commented:
> yes, we can have improve the test later, also maybe add some es6 code to see if it's preserved etc ...

Definitely. It's worth thinking about on what to test.
